### PR TITLE
Refactor merge to support partial update

### PR DESCRIPTION
**Pull request by chenjian2664**
## Description

Currently the `update` implementation of MERGE update updates the whole row, this pr is going to update the engine merge impl to support partial update.
The main idea is to keep the `merge case_number` of the `UpdateCase` in the output of the `MergeProcessNode`,(this value already kept in the `merge_row` symbol that in the sourceNode of the `MergeProcessNode`, so only needs to project it), and then in the `MergeWriterNode`, the `updateSink` can perform the update according to the update case.

The partial update also needs to know the mapping of the case number with the relevant columns, thus the `ConnectorMetadata#beginMerge` is refactored to hold the info.

Here use the Phoenix connector as a example to show how it works.

## Additional context and related issues

Prerequisites for https://github.com/trinodb/trino/pull/23034

## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
(x) Release notes are required. Please propose a release note for me.
() Release notes are required, with the following suggested text:

```markdown
## Section
* Support partial update in engine.
* Support partial update in Phoenix.
```

### CODE REVIEW

**Review by djsstarburst — COMMENTED** [review]
This looked pretty good to me.  Thanks for undertaking this work!

The PR needs Phoenix tests that show that only updated columns are changed.  The tests need to cover cases where a single column is updated in a single row and in multiple rows; where non-overlapping sets of columns are updated in different cases; and tests that include UPDATE as well as INSERT and DELETE cases.

We also need @electrum to take a look.

File: core/trino-main/src/main/java/io/trino/metadata/Metadata.java [hunk 1]
```diff
@@ -453,7 +454,7 @@ Optional<ConnectorOutputMetadata> finishRefreshMaterializedView(
     /**
      * Begin merge query
      */
-    MergeHandle beginMerge(Session session, TableHandle tableHandle);
+    MergeHandle beginMerge(Session session, TableHandle tableHandle, Multimap<Integer, ColumnHandle> updateCaseColumnHandles);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/metadata/Metadata.java:457]: What is meaning of the key? I think this could be explained in javadoc.

Comment (chenjian2664) [reply core/trino-main/src/main/java/io/trino/metadata/Metadata.java:458]: I leave the `The merge update case number with the relevant columns mapping`, but still want to get your suggestions, could you please help to phrase the sentence?  cc @kokosing @djsstarburst 
The mapping here is the `MergeUpdate` case number to it's setExpressions assignment target, then update would known what are the target columns. Take the a example:
```
MERGE INTO t USING s 
ON t.pk = s.pk 
WHEN MATCHED AND t.c1 > 0 THEN DELETE // case 0
WHEN MATCHED AND t.c2 < 10 THEN UPDATE SET t.c2 = 10, t.c3 = 10 // case 1
WHEN MATCHED AND t.c2 < 100 THEN UPDATE SET t.c4 = 100, t.c3 = 100 // case 2
WHEN MATCHED THEN UPDATE SET t.c5 = s.c5 + 10 // case 3
WHEN NOT MATCHED THEN INSERT.... // case 4
``` 
the map value would be like: {1: [c2, c3], 2: [c4, c3], 3: c5}

**Comment by chenjian2664** [comment]
Flaky https://github.com/trinodb/trino/actions/runs/11821799275/job/32937454502?pr=24075#step:5:2219

**Comment by chenjian2664** [comment]
https://github.com/trinodb/trino/actions/runs/11829608194/job/32961756024?pr=24075#step:5:2343

**Comment by chenjian2664** [comment]
> https://github.com/trinodb/trino/actions/runs/11829608194/job/32961756024?pr=24075#step:5:2343

https://github.com/trinodb/trino/issues/24131

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java [hunk 2]
```diff
@@ -796,6 +799,283 @@ public void testExecuteProcedure()
         }
     }
 
+    @Test
+    public void testMergeUpdatePartial()
```
Comment (djsstarburst) [inline plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:803]: This is a very nice test, and just what I requested!

**Review by djsstarburst — APPROVED** [review]
I'm approving this PR, but we still need @electrum to take a look.  I'll ping him now about it.

File: plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java (hunk 2 above)
Comment (chenjian2664) [reply plugin/trino-phoenix5/src/test/java/io/trino/plugin/phoenix5/TestPhoenixConnectorTest.java:803]: Glad to hear! Thanks for the guidance!

File: core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java [hunk 3]
```diff
@@ -193,11 +198,14 @@ private void addInsertRow(PageBuilder pageBuilder, List<Block> fields, int posit
         // Add the operation column == insert
         TINYINT.writeLong(pageBuilder.getBlockBuilder(dataColumnChannels.size()), causedByUpdate ? UPDATE_INSERT_OPERATION_NUMBER : INSERT_OPERATION_NUMBER);
 
+        // Add the dummy case number, delete and insert won't use it
+        INTEGER.writeLong(pageBuilder.getBlockBuilder(dataColumnChannels.size() + 1), 0);
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/operator/DeleteAndInsertMergeProcessor.java:202]: In other places you do:
```
Update only have one update case number which default is 0
```

So let's pick different number. Maybe `-1` to make it clear it is incorrect and so it should not be used.

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -3645,12 +3648,31 @@ else if (operation instanceof MergeInsert && caseColumnNames.isEmpty()) {
             analysis.setUpdateTarget(targetTableHandle.catalogHandle().getVersion(), tableName, Optional.of(table), Optional.of(updatedColumns));
             List<List<ColumnHandle>> mergeCaseColumnHandles = buildCaseColumnLists(merge, dataColumnSchemas, allColumnHandles);
 
-            createMergeAnalysis(table, targetTableHandle, tableSchema, targetTableScope, joinScope, mergeCaseColumnHandles);
+            checkArgument(mergeCaseColumnHandles.size() == merge.getMergeCases().size(),
+                    "Unexpected mergeCaseColumnHandles size: %s with merge cases size: %s", mergeCaseColumnHandles.size(), merge.getMergeCases().size());
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3652]:
```
            checkArgument(
                    mergeCaseColumnHandles.size() == merge.getMergeCases().size(),
                    "Unexpected mergeCaseColumnHandles size: %s with merge cases size: %s", mergeCaseColumnHandles.size(), merge.getMergeCases().size());
```

File: core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java [hunk 5]
```diff
@@ -3645,12 +3648,31 @@ else if (operation instanceof MergeInsert && caseColumnNames.isEmpty()) {
...
+            ImmutableMultimap.Builder<Integer, ColumnHandle> updateCaseColumnHandlesBuilder = ImmutableMultimap.builder();
```
Comment (kokosing) [inline core/trino-main/src/main/java/io/trino/sql/analyzer/StatementAnalyzer.java:3653]: s/updateCaseColumnHandlesBuilder/updateCaseColumnHandles

**Review by kokosing — COMMENTED** [review]
Reviewed initially first commit.

**Review by electrum — APPROVED** [review]

**Comment by electrum** [comment]
Thanks!

**Pull request merged**